### PR TITLE
[camera] Limit video length with maxVideoDuration on startVideoRecording

### PR DESCRIPTION
**Pull request by danielroek**
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

Adds implementation for defining a maximum video duration.

*List which issues are fixed by this PR. You must list at least one issue.*

TODO

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

TODO

## Pre-launch Checklist

- [ ] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 1]
```diff
@@ -170,16 +171,25 @@ private void initFps(CameraCharacteristics cameraCharacteristics) {
...
+                      .setMediaOrientation(getMediaOrientation())
+                      .setMaxVideoDuration(maxVideoDuration)
+                      .build();
+    } else {
+      mediaRecorder =
+              new MediaRecorderBuilder(recordingProfile, outputFilePath)
+                      .setEnableAudio(enableAudio)
+                      .setMediaOrientation(getMediaOrientation())
+                      .build();
+    }
```
Comment (danielroek) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:192]: Simplify?

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 2]
```diff
@@ -589,10 +599,26 @@ public void startVideoRecording(Result result) {
...
+     if(maxVideoDuration != null) {
+       mediaRecorder.setOnInfoListener((mr, what, extra) -> {
+         if (what == MEDIA_RECORDER_INFO_MAX_DURATION_REACHED) {
+           try {
+             dartMessenger.sendTimeLimitReachedEvent(videoRecordingFile.getAbsolutePath());
+             cameraCaptureSession.abortCaptures();
+             mediaRecorder.stop();
+             recordingVideo = false;
+             videoRecordingFile = null;
+             startPreview();
```
Comment (danielroek) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:615]: Maybe combine this with the code in stopVideoRecording()?

File: packages/camera/camera/example/lib/main.dart [hunk 3]
```diff
@@ -606,7 +606,14 @@ class _CameraExampleHomeState extends State<CameraExampleHome>
...
     try {
-      await controller.startVideoRecording();
+      await controller.startVideoRecording(
+          maxVideoDuration: const Duration(milliseconds: 5000),
+      );
+      debugPrint('recording started');
+      controller.onCameraTimeLimitReachedEvent(onCameraTimeLimitReached: (file) {
+        debugPrint('onCameraTimeLimitReached ${file.path}');
+      });
+      debugPrint('listening');
```
Comment (danielroek) [inline packages/camera/camera/example/lib/main.dart:616]: Remove here, add to README

File: packages/camera/camera/lib/src/camera_controller.dart [hunk 4]
```diff
@@ -482,6 +485,32 @@ class CameraController extends ValueNotifier<CameraValue> {
...
 
+  /// TODO: Documentation
+  void onCameraTimeLimitReachedEvent({onCameraTimeLimitReached}) {
+    if (!value.isInitialized || _isDisposed) {
+      throw CameraException(
+        'Uninitialized CameraController',
+        'cameraTimeLimitReachedEventStream was called on uninitialized CameraController',
+      );
+    }
+    if (!value.isRecordingVideo) {
```
Comment (danielroek) [inline packages/camera/camera/lib/src/camera_controller.dart:496]: Is this correct behaviour? Should we only listen after starting the recording? Or should it be possible beforehand?

File: packages/camera/camera/lib/src/camera_controller.dart [hunk 5]
```diff
@@ -482,6 +485,32 @@ class CameraController extends ValueNotifier<CameraValue> {
...
+      throw CameraException(
+        'No video is recording',
+        'cameraTimeLimitReachedEventStream was called when no video is recording.',
+      );
+    }
+    debugPrint('ping');
+
+    CameraPlatform.instance.onCameraTimeLimitReached(_cameraId).listen((event) {
```
Comment (danielroek) [inline packages/camera/camera/lib/src/camera_controller.dart:504]: Doesn't get closed after receiving an event, onCameraTimeLimitReached gets printed multiple times in 2nd+ recording.

File: packages/camera/camera/pubspec.yaml [hunk 6]
```diff
@@ -8,7 +8,8 @@ homepage: https://github.com/flutter/plugins/tree/master/packages/camera/camera
 dependencies:
   flutter:
     sdk: flutter
-  camera_platform_interface: ^1.2.0
+  camera_platform_interface: #^1.5.0
+    path: ../camera_platform_interface
```
Comment (danielroek) [inline packages/camera/camera/pubspec.yaml:12]: Remove path

File: packages/camera/camera/lib/src/camera_controller.dart [hunk 7]
```diff
@@ -482,6 +485,30 @@ class CameraController extends ValueNotifier<CameraValue> {
...
+    }
+
+    CameraPlatform.instance
+        .onCameraTimeLimitReached(_cameraId)
+        .first
+        .then((event) {
+      value = value.copyWith(isRecordingVideo: false);
+      onCameraTimeLimitReached(event.path);
+    });
+  }
```
Comment (BeMacized) [inline packages/camera/camera/lib/src/camera_controller.dart:510]: One issue I see with this, is that if the user sets a `maxVideoDuration`, but never calls `onCameraTimeLimitReachedEvent` to handle the callback, `value` will never be updated, causing `isRecordingVideo to still report `true`, while it has in fact already stopped. 

Ideally, the camera controller should always be listening for `CameraPlatform.instance.onCameraTimeLimitReached` to update the CameraValue, even when the user has not provided a callback, so that the state is always correct. 

Additionally, I think I would personally prefer seeing a method that returns a stream of these events (Like how the platform interface exposes events) rather than a method that allows for setting a callback, as that would imo offer a bit more flexibility. What do you think?

File: packages/camera/camera_platform_interface/lib/src/events/camera_event.dart [hunk 8]
```diff
@@ -235,3 +235,33 @@ class CameraErrorEvent extends CameraEvent {
   @override
   int get hashCode => super.hashCode ^ description.hashCode;
 }
+
+class CameraTimeLimitReachedEvent extends CameraEvent {
+  final XFile path;
+
```
Comment (BeMacized) [inline packages/camera/camera_platform_interface/lib/src/events/camera_event.dart:241]: Should the event maybe also include the time limit that was reached? I have a feeling there might be a use case for this.

File: packages/camera/camera_platform_interface/lib/src/platform_interface/camera_platform.dart [hunk 9]
```diff
@@ -108,6 +108,11 @@ abstract class CameraPlatform extends PlatformInterface {
     throw UnimplementedError('unlockCaptureOrientation() is not implemented.');
   }
 
+  @override
+  Stream<CameraTimeLimitReachedEvent> onCameraTimeLimitReached(int cameraId) {
+    throw UnimplementedError('onCameraTimeLimitReached() is not implemented.');
+  }
```
Comment (BeMacized) [inline packages/camera/camera_platform_interface/lib/src/platform_interface/camera_platform.dart:114]: This change to the platform interface should probably be split off to a separate PR, so that it can be published to pub.dev before the rest is merged to master.

**Review by BeMacized — COMMENTED** [review]
Overall this already looks pretty good! 😄 
I've left a few comments below.

Additionally:
- Unit tests are still missing for your changes.
- iOS implementations are still missing.

File: packages/camera/camera_platform_interface/lib/src/events/camera_event.dart (hunk 8 above)
Comment (mvanbeusekom) [reply packages/camera/camera_platform_interface/lib/src/events/camera_event.dart:241]: I agree, this would be helpful information to have around.

File: packages/camera/camera/lib/src/camera_controller.dart (hunk 7 above)
Comment (mvanbeusekom) [reply packages/camera/camera/lib/src/camera_controller.dart:510]: I agree, this would be easier for users to manage if we can use this as a `Stream<XFile>` and simply listen for a file to arrive. 

We can even extend the `stopVideoRecording` method to add the final video file to the same stream. This way developers simply listen to the stream to get informed that a new video file is available (no matter if the end-user pressed the stop button or if the max time limit has been reached). Of course this would mean the name of the event and the stream should be updated a to reflect this change. Something like:

```dart
Stream<VideoRecordedEvent> onVideoRecorded(int cameraId); 
```

File: packages/camera/camera_platform_interface/lib/src/events/camera_event.dart [hunk 10]
```diff
@@ -235,3 +235,33 @@ class CameraErrorEvent extends CameraEvent {
   @override
   int get hashCode => super.hashCode ^ description.hashCode;
 }
+
+class CameraTimeLimitReachedEvent extends CameraEvent {
+  final XFile path;
```
Comment (mvanbeusekom) [inline packages/camera/camera_platform_interface/lib/src/events/camera_event.dart:240]: nit: Rename to `file` instead of `path`:
```suggestion
  final XFile file;
```

File: packages/camera/camera_platform_interface/lib/src/method_channel/method_channel_camera.dart [hunk 11]
```diff
@@ -162,6 +162,11 @@ class MethodChannelCamera extends CameraPlatform {
         .whereType<DeviceOrientationChangedEvent>();
   }
 
+  @override
+  Stream<CameraTimeLimitReachedEvent> onCameraTimeLimitReached(int cameraId) {
```
Comment (mvanbeusekom) [inline packages/camera/camera_platform_interface/lib/src/method_channel/method_channel_camera.dart:166]: I think it would be better to rename this to `onVideoRecordingTimeLimitReached`, so it is clear this is directly related to the video recording feature. Of course if you accept my earlier comment about also emitting the recorded video on this stream when the user calls `stopVideoRecording` we should rename it to something more generic like `onVideoRecorded`.

File: packages/camera/camera_platform_interface/lib/src/platform_interface/camera_platform.dart (hunk 9 above)
Comment (mvanbeusekom) [reply packages/camera/camera_platform_interface/lib/src/platform_interface/camera_platform.dart:114]: I think it would be better to rename this to `onVideoRecordingTimeLimitReached`, so it is clear this is directly related to the video recording feature. Of course if you accept my earlier comment about also emitting the recorded video on this stream when the user calls stopVideoRecording we should rename it to something more generic like `onVideoRecorded`.

File: packages/camera/camera/README.md [hunk 12]
```diff
@@ -66,6 +66,33 @@ As of version [0.5.0](https://github.com/flutter/plugins/blob/master/packages/ca
...
+        );
+        controller.onCameraTimeLimitReachedEvent(onCameraTimeLimitReached: (XFile file) {
+          //Handle the XFile
+        });
+      } on CameraException catch (e) {
+        _showCameraException(e);
+        return;
+      }
+    }
+```
```
Comment (mvanbeusekom) [inline packages/camera/camera/README.md:95]: Really like that you didn't forget to updated the documentation in the README.md. But this should probably also be updated according the the discussion @BeMacized started on using a stream instead of registering a callback method.

**Review by mvanbeusekom — CHANGES_REQUESTED** [review]

File: packages/camera/camera/lib/src/camera_controller.dart (hunk 7 above)
Comment (danielroek) [reply packages/camera/camera/lib/src/camera_controller.dart:510]: This might also solve possible issues with recording video's and stopping them for unusual reasons, like full storage

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 13]
```diff
@@ -601,6 +619,18 @@ public void startVideoRecording(Result result) {
...
+  public void resetCaptureSession() throws CameraAccessException {
+    try {
+      cameraCaptureSession.abortCaptures();
+      mediaRecorder.stop();
+    } catch (CameraAccessException | IllegalStateException e) {
+      // Ignore exceptions and try to continue (changes are camera session already aborted capture)
+    }
+
+    mediaRecorder.reset();
+    startPreview();
```
[the remark this answers is not in the document]
Comment (danielroek) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:655]: Removed the CameraException from the catch-clause. I'm going to leave the finally statement, as it does add the risk of making future problems harder to debug.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 14]
```diff
@@ -610,15 +640,8 @@ public void stopVideoRecording(@NonNull final Result result) {
...
-      } catch (CameraAccessException | IllegalStateException e) {
-        // Ignore exceptions and try to continue (changes are camera session already aborted capture)
-      }
+      resetCaptureSession();
 
-      mediaRecorder.reset();
-      startPreview();
       result.success(videoRecordingFile.getAbsolutePath());
       videoRecordingFile = null;
     } catch (CameraAccessException | IllegalStateException e) {
```
[the remark this answers is not in the document]
Comment (danielroek) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:671]: That's right

**Comment by kiaxseventh** [comment]
add a full example please
